File: 2021/07/31/valid_palindrome.py
"""
https://leetcode.com/problems/valid-palindrome
Author: Andrew Jarombek
Date: 7/31/2021
"""

import re


def is_palindrome(s: str) -> bool:
    pattern = re.compile(r"[^a-zA-Z0-9]+")
    lowercase_str = re.sub(pattern, '', s).lower()

    for i in range(len(lowercase_str) // 2):
        if lowercase_str[i] != lowercase_str[-1 - i]:
            return False

    return True


if __name__ == '__main__':
    assert is_palindrome('racecar')
    assert not is_palindrome('andy')

File: 2022/11/18/cpp/validPalindrome.cpp
// https://leetcode.com/problems/valid-palindrome/
// Author: Andrew Jarombek
// Date: 11/18/2022

#include <cassert>
#include <vector>
#include <string>

using namespace std;

// I hope to be a calming and relaxing presence, and to be extremely understanding and gentle towards any challenges
// you are facing.  I still hope for and haven’t given up on you trusting yourself with love and trusting me with love,
// if that is something you still want.

bool isPalindrome(string s) {
    int start = 0;
    int end = s.size() - 1;
    bool result = true;

    while (start < end) {
        char a = tolower(s[start]);

        while (start < end && (a < 'a' || a > 'z') && (a < '0' || a > '9')) {
            start++;
            a = tolower(s[start]);
        }

        char b = tolower(s[end]);

        while (start < end && (b < 'a' || b > 'z') && (b < '0' || b > '9')) {
            end--;
            b = tolower(s[end]);
        }

        if (a != b)
            return false;

        start++;
        end--;
    }

    return true;
}

int main() {
    assert(!isPalindrome("race a car"));
    assert(isPalindrome("race car"));
}